subprojects/file-collections/src/main/java/org/gradle/api/internal/file/AbstractImmutableFileAccessPermissions.java

```diff
@@ -0,0 +1,55 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.internal.file;
+
+import org.gradle.api.file.ImmutableFileAccessPermissions;
+import org.gradle.api.internal.provider.Providers;
+import org.gradle.api.provider.Provider;
+
+public abstract class AbstractImmutableFileAccessPermissions implements ImmutableFileAccessPermissions {
+
+    @Override
+    public Provider<Integer> toUnixNumeric() {
+        return Providers.changing(() ->
+            toUnixNumericPermissions(
+                ((AbstractImmutableFileAccessPermission) getUser()).toUnixNumeric(),
+                ((AbstractImmutableFileAccessPermission) getGroup()).toUnixNumeric(),
+                ((AbstractImmutableFileAccessPermission) getOther()).toUnixNumeric()
+            )
+        );
+    }
```

Comment:
Following up on the provider chain breakage in `AbstractImmutableFileAccessPermission` above, here's how I think this should be implemented:

```suggestion
    public Provider<Integer> toUnixNumeric() {
        AbstractImmutableFileAccessPermission user = (AbstractImmutableFileAccessPermission) getUser();
        AbstractImmutableFileAccessPermission group = (AbstractImmutableFileAccessPermission) getGroup();
        AbstractImmutableFileAccessPermission other = (AbstractImmutableFileAccessPermission) getOther();
        return user.toUnixNumeric().map(u -> 64 * u)
            .zip(group.toUnixNumeric().map(g -> 8 * g), Integer::sum)
            .zip(other.toUnixNumeric(), Integer::sum);
    }
```